app/src/main/java/com/javahelp/frontend/activity/ProviderRegistrationActivity.java

```diff
@@ -25,14 +50,109 @@ protected void onCreate(Bundle savedInstanceState) {
+            providerRegistrationViewmodel.setRegisterError("Please enter your personal information");
+            return;
+        }
+    }
+
+    /**
+     * Updates this {@link RegisterResult} to reflect the passed register result
+     *
+     * @param registerResult {@link Optional} {@link RegisterResult} to update on
+     */
+    private void updateOnRegisterResult(Optional<RegisterResult> registerResult) {
+        if(registerResult.isPresent()){
+            updateOnPresentRegisterResult(registerResult.get());
+        } else {
+            binding.registerErrorText.setVisibility(View.GONE);
+        }
+    }
+
+    /**
+     * Updates the UI based on a {@link RegisterResult}. This is called with non-null register results, meaning
+     * a registration has been attempted, and has either succeeded or failed. This is unlike updateOnRegisterResult
+     * which is called with an {@link Optional} {@link RegisterResult}, and calls this method if that
+     * {@link Optional} is present.
+     *
+     * @param result {@link RegisterResult} of login attempt
+     */
+    private void updateOnPresentRegisterResult(RegisterResult result) {
+        if (result.isSuccess()) {
+            binding.registerErrorText.setText("Register successful");
+            storeCredentials(result.getUser(), result.getToken());
+            startActivity(new Intent(ProviderRegistrationActivity.this, FrontPageActivity.class));
+        } else {
+            binding.registerErrorText.setText(result.getErrorMessage());
+        }
+        binding.registerErrorText.setVisibility(View.VISIBLE);
+    }
 
+    /**
+     * Stores the accepted credentials
+     *
+     * @param user  {@link User} to store
+     * @param token {@link Token} to store
+     */
+    private void storeCredentials(User user, Token token) {
+        SharedPreferencesAuthInformationProvider credentialStore =
+                new SharedPreferencesAuthInformationProvider(this);
+        try {
+            credentialStore.setTokenString(token.getToken());
+            credentialStore.setUserID(user.getStringID());
+        } catch (NoSuchPaddingException | IllegalBlockSizeException | NoSuchAlgorithmException | BadPaddingException | InvalidKeyException ignored) {
+            Toast.makeText(this, "Error Storing Credentials", Toast.LENGTH_LONG);
+        }
     }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        switch (requestCode) {
+            case REQUEST_INTERNET_REGISTER:
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    providerRegistrationViewmodel.attemptRegister();
+                } else {
+                    Toast.makeText(this, "Internet access denied, cannot log in", Toast.LENGTH_LONG);
+                }
+                return;
```

Comment:
I personally think this line 154 is not necessary.

Reply:
It is unnecessary, but we may want to add more cases in the future, in which case we will require either a return or a break